project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/emf/EMFGModelFactory.java

```diff
@@ -0,0 +1,98 @@
+/********************************************************************************
+ * Copyright (c) 2022 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * https://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+package org.eclipse.glsp.example.javaemf.emf;
+
+import com.google.inject.Inject;
+
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.util.EcoreUtil;
+import org.eclipse.emfcloud.modelserver.glsp.notation.Diagram;
+import org.eclipse.emfcloud.modelserver.glsp.notation.Shape;
+import org.eclipse.glsp.graph.GGraph;
+import org.eclipse.glsp.graph.GModelRoot;
+import org.eclipse.glsp.graph.builder.impl.GGraphBuilder;
+import org.eclipse.glsp.graph.builder.impl.GNodeBuilder;
+import org.eclipse.glsp.graph.util.GraphUtil;
+import org.eclipse.glsp.server.features.core.model.GModelFactory;
+import org.eclipse.glsp.server.model.GModelState;
+
+/**
+ * A graph model factory produces a graph model from the contained models of the EMF model state {@link EMFGModelState}.
+ * It expects a semantic as well as an notation model to create to the GModel.
+ * Subclasses need to provide a diagram factory that translates the semantic and notation model information to a GModel.
+ */
+public abstract class EMFGModelFactory implements GModelFactory {
+
+    @Inject
+    protected GModelState gModelState;
```

Comment:
I think you could inject the type `EMFGModelState` here directly and avoid having to cast later on.

Reply:
I injected the specific ModelState type where possible (i.e. the `TaskListGModelState`). For the others (`EMFGModelState` and `EMFNotationGModelState`) I went with the casting approach for very few occurrences as we would have different instances (as we only bind `TaskListGModelState` directly) which causes problems.